### PR TITLE
Feature clone dataset

### PR DESCRIPTION
**Pull request by MashB**
Dataset shallow copy user story 
https://dev.azure.com/TDEI-UW/TDEI/_workitems/edit/994/

Functional Flow 
https://miro.com/app/board/uXjVPU6Y7lY=/?moveToWidget=3458764588688202138&cot=14

### CODE REVIEW

**Review by susrisha — CHANGES_REQUESTED** [review]
minor changes

File: src/controller/general-controller.ts [hunk 1]
```diff
@@ -42,21 +43,70 @@ class GeneralController implements IController {
...
+                    } else if (datasetRecord.data_type === TDEIDataType.flex) {
+                        authorize([TDEIRole["tdei-admin"], TDEIRole.poc, TDEIRole.flex_data_generator])(req, res, next);
+                    } else if (datasetRecord.data_type === TDEIDataType.pathways) {
+                        authorize([TDEIRole["tdei-admin"], TDEIRole.poc, TDEIRole.pathways_data_generator])(req, res, next);
+                    }
+                } catch (error) {
+                    next(error);
+                }
+            }, this.editMetadata); // edit Metadata request
+        this.router.post(`${this.path}/dataset/clone/:tdei_dataset_id/:tdei_project_group_id/:tdei_service_id`, metadataUpload.single('file'), authenticate, async (req, res, next) => {
```
[the remark this answers is not in the document]
Comment (MashB) [reply src/controller/general-controller.ts:60]: Current should be good, it is read as clone {dataset} to {projectid} / {serviceid}

File: src/service/storage-service.ts [hunk 2]
```diff
@@ -4,12 +4,21 @@
...
 import { Core } from "nodets-ms-core";
+import { FileEntity } from "nodets-ms-core/lib/core/storage";
 
 class StorageService {
 
     generateRandomUUID(): string {
-        const randomUID = randomUUID().toString().replace(/-/g, ''); // Take out the - from UID
-        return randomUID;
+        // const randomUID = randomUUID().toString().replace(/-/g, ''); // Take out the - from UID
+        return randomUUID().toString();
```
[the remark this answers is not in the document]
Comment (MashB) [reply src/service/storage-service.ts:13]: Earlier we removed hypen for readability, with cloning we are removing this as DB generates UUID with hyphen.

File: src/service/storage-service.ts [hunk 3]
```diff
@@ -4,12 +4,21 @@
...
+    }
+
+    getStorageFileNameFromUrl(fullUrl: string): string {
```
[the remark this answers is not in the document]
Comment (MashB) [reply src/service/storage-service.ts:16]: done

File: src/service/storage-service.ts [hunk 4]
```diff
@@ -61,6 +70,29 @@ class StorageService {
...
+        let clonedFileEntity = client?.cloneFile(decodeURIComponent(fileUrl), destinationContainerName, destinationFilePath);
+        return clonedFileEntity;
+    }
+
+    /**
+     * Deletes a file from the storage
+     * @param fileUrl file url
+     */
+    async deleteFile(fileUrl: string) {
+        const client = Core.getStorageClient();
```
[the remark this answers is not in the document]
Comment (MashB) [reply src/service/storage-service.ts:92]: done

File: src/database/dynamic-update-query.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    const values = Object.values(data);
+    const setString = keys.map((key, i) => `${key} = $${i + 1}`).join(', ');
+
+    const whereKeys = Array.from(where.keys());
+    const whereValues = Array.from(where.values());
+    const whereString = whereKeys.map((key, i) => `${key} = $${values.length + i + 1}`).join(' AND ');
+
+    values.push(...whereValues);
+
+    const query = {
```
Comment (ayydevara) [inline src/database/dynamic-update-query.ts:12]: there is a possibility of mal-formed query if where map is empty or null, you should add checks to handle those cases

File: src/database/dynamic-update-query.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+export function buildUpdateQuery(table: string, data: any, where: Map<string, string>, returnUpdatedRow: boolean = false) {
+    const keys = Object.keys(data);
```
Comment (ayydevara) [inline src/database/dynamic-update-query.ts:2]: you can probably move all of this to a builder class `QueryCriteria`. This should be fine too, make sure you handle the corner cases and base cases

Comment (ayydevara) [reply src/database/dynamic-update-query.ts:2]: something like this...

```ts
const criteria = new QueryCriteria()
  .where('id', 1)
  .where('status', 'active', '!=')
  .set('name', 'New Name')
  .set('updated_at', new Date());
  
  console.log(criteria.build());
  
  ```

File: src/service/storage-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ayydevara) [reply src/service/storage-service.ts:24]: do we want `-` or no `-` in uuid?

File: src/service/storage-service.ts [hunk 7]
```diff
@@ -2,14 +2,39 @@
...
+    /**
+     * Gets the storage file name from the full URL
+     * @param fullUrl full URL of the file
+     * @returns string with the file name
+     */
+    getStorageFileNameFromUrl(fullUrl: string): string {
+        let url = new URL(fullUrl);
+        let pathname = url.pathname;
+        let filenameWithExtension = pathname.split("/").pop();
+        // let filenameWithoutExtension = filenameWithExtension!.split(".").slice(0, -1).join(".");
```
Comment (ayydevara) [inline src/service/storage-service.ts:36]: remove comments

File: src/service/tdei-core-service.ts [hunk 8]
```diff
@@ -361,8 +364,234 @@ class TdeiCoreService implements ITdeiCoreService {
...
+            dest_metadata_upload_entity: undefined,
+            dest_osm_upload_entity: undefined,
+            new_tdei_dataset_id: "",
+            dest_dataset_download_entity: undefined,
+            dest_osm_download_entity: undefined
+        };
+
+        try {
+            //Get the dataset details to be cloned
+            let dataset_to_be_clone = await this.getDatasetDetailsById(datasetCloneRequestObject.tdei_dataset_id);
```
Comment (ayydevara) [inline src/service/tdei-core-service.ts:390]: do we need comments?

Comment (ayydevara) [reply src/service/tdei-core-service.ts:390]: **Comments:**
Remove comments that explain straightforward logic. 
function names and clean code make the comments redundant.

The self-explanatory nature of the code eliminates the need for excessive commenting.
This approach improves readability and maintainability by using clear and descriptive function names and minimizing unnecessary comments.

Comment (MashB) [reply src/service/tdei-core-service.ts:390]: added as code comments for readability.

File: src/service/storage-service.ts (hunk 7 above)
Comment (MashB) [reply src/service/storage-service.ts:36]: Done

File: src/database/dynamic-update-query.ts (hunk 6 above)
Comment (MashB) [reply src/database/dynamic-update-query.ts]: Taken care, written unit test to validate this functionality. utility.test.ts

File: src/database/dynamic-update-query.ts (hunk 5 above)
Comment (MashB) [reply src/database/dynamic-update-query.ts]: Taken care

File: src/service/storage-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MashB) [reply src/service/storage-service.ts:23]: Earlier we removed hyphen for readability, now to be align with DB UUId generator we need hyphen. (clone happens from script now)

File: src/service/tdei-core-service.ts (hunk 8 above)
Comment (MashB) [reply src/service/tdei-core-service.ts:389]: Removed comments.  function names & variables are self descriptive.

File: src/database/dynamic-update-query.ts (hunk 6 above)
Comment (MashB) [reply src/database/dynamic-update-query.ts]: Added builder pattern

File: src/service/storage-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MashB) [reply src/service/storage-service.ts:23]: We want hyphen.

**Review by susrisha — APPROVED** [review]

**Pull request merged**